src/stateManager/StateManager.ts

```diff
@@ -0,0 +1,827 @@
+    public async onBlockConfirmation(
+        signedBlock: SignedBlockStruct,
+        confirmationSignature: BytesLike
+    ): Promise<ExecutionFlags> {
+        let finalExecutionFlag: ExecutionFlags = ExecutionFlags.SUCCESS; // Default to SUCCESS
+
+        try {
+            const result = await this.validateBlockConfirmation(
+                signedBlock,
+                confirmationSignature
+            );
+            finalExecutionFlag = result.flag;
+
+            if (result.success) {
+                const block = EvmUtils.decodeBlock(signedBlock.encodedBlock);
+                this.agreementManager.confirmBlock(
+                    block,
+                    confirmationSignature as SignatureLike
+                );
+            }
+
+            return finalExecutionFlag;
+        } finally {
+            if (finalExecutionFlag === undefined) {
+                throw new Error(
+                    "StateManager - onBlockConfirmation - Internal Error - flag undefined"
+                );
+            }
+
+            await this.processConfirmationDecision(
+                signedBlock,
+                confirmationSignature as SignatureLike,
+                finalExecutionFlag
+            );
+        }
+    }
+
+    //Aplies a transaction to the state machine and returns the encoded state with a success callback
+    public async applyTransaction(transaction: TransactionStruct): Promise<{
+        success: boolean;
+        encodedState: string;
+        successCallback: () => void;
+    }> {
+        let { success, successCallback } =
+            await this.stateMachine.stateTransition(transaction);
+        return {
+            success,
+            encodedState: await this.stateMachine.getState(),
+            successCallback
+        };
+    }
+
+    // Used when authoring a block - Executes the transaction and returns a signed block
+    public async playTransaction(
+        tx: TransactionStruct
+    ): Promise<SignedBlockStruct> {
+        await this.mutex.lock();
+
+        try {
+            console.log("Play Transaction", this.getForkCnt());
+            this.ensureChannelIsOpen();
+            await this.ensureItIsMyTurn();
+            this.adjustTimestampIfNeeded(tx);
+
+            const { previousStateHash, encodedState, successCallback } =
+                await this.applyTransactionOrThrow(tx);
+
+            const block = await this.createBlock(tx, previousStateHash);
+            const signedBlock = await this.signBlock(block);
+
+            this.agreementManager.addBlock(
+                block,
+                signedBlock.signature as SignatureLike,
+                encodedState
+            );
+
+            successCallback();
+            await this.onSuccessCommon();
+
+            this.scheduleTask(
+                () => this.maybePostBlockOnChain(block, signedBlock),
+                this.timeConfig.agreementTime * 1000,
+                "maybePostBlockOnChain"
+            );
```

Comment:
@daiagi you didn't need to close the PR - there are good things in it like the playTransaction refactor - everything here is reusable 

The Task scheduling is another useful thing since it wraps the setTimeout in a reusable way that's more clear.

Partitioning this file into a stateManager directory and breaking parts into files like processConfirmation/ExecutionDecision is also good

Any refactor that simplifies, wraps and makes things reusable without adding complexity is good.

Reply:
ok. re-open
 you are right that the validator attempt has become useless
 i'll remove that, and use StateManger.<validator>  that returns plain boolean instead